stko/optimizers/xtb.py

```diff
@@ -0,0 +1,509 @@
+
+    def _run_xtb(self, xyz, out_file):
+        """
+        Run GFN-xTB.
+
+        Parameters
+        ----------
+        xyz : :class:`str`
+            The name of the input structure ``.xyz`` file.
+
+        out_file : :class:`str`
+            The name of output file with xTB results.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        # Modify the memory limit.
+        if self._unlimited_memory:
+            memory = 'ulimit -s unlimited ;'
+        else:
+            memory = ''
+
+        # Set optimization level and type.
+        if self._calculate_hessian:
+            # Do optimization and check hessian.
+            optimization = f'--ohess {self._opt_level}'
+        else:
+            # Do optimization.
+            optimization = f'--opt {self._opt_level}'
+
+        if self._solvent is not None:
+            solvent = f'--gbsa {self._solvent} {self._solvent_grid}'
+        else:
+            solvent = ''
+
+        cmd = (
+            f'{memory} {self._xtb_path} {xyz} '
+            f'--gfn {self._gfn_version} '
+            f'{optimization} --parallel {self._num_cores} '
+            f'--etemp {self._electronic_temperature} '
+            f'{solvent} --chrg {self._charge} '
+            f'--uhf {self._num_unpaired_electrons}'
+        )
+
+        with open(out_file, 'w') as f:
+            # Note that sp.call will hold the program until completion
+            # of the calculation.
+            sp.call(
+                cmd,
+                stdin=sp.PIPE,
+                stdout=f,
+                stderr=sp.PIPE,
+                # Shell is required to run complex arguments.
+                shell=True
+            )
+
+    def _run_optimizations(self, mol):
+        """
+        Run loop of optimizations on `mol` using xTB.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        mol : :class:`.Molecule`
+            The optimized molecule.
+
+        opt_complete : :class:`bool`
+            Returns ``True`` if the calculation is complete and
+            ``False`` if the calculation is incomplete.
+
+        """
+
+        for run in range(self._max_runs):
+            xyz = f'input_structure_{run+1}.xyz'
+            out_file = f'optimization_{run+1}.output'
+            mol.write(xyz)
+            self._run_xtb(xyz=xyz, out_file=out_file)
+            # Check if the optimization is complete.
+            coord_file = 'xtbhess.coord'
+            coord_exists = os.path.exists(coord_file)
+            output_xyz = 'xtbopt.xyz'
+            opt_complete = self._is_complete(out_file)
+            if not opt_complete:
+                if coord_exists:
+                    # The calculation is incomplete.
+                    # Update mol from xtbhess.coord and continue.
+                    mol = mol.with_structure_from_file(coord_file)
```

Comment:
Best if a new variable i.e. new_mol is made instead of overwriting the stk object

Reply:
I disagree considering that would confuse the purpose of the function (to optimize mol, no reason it should produce the old mol in any case [failure or optimisation])

Reply:
But it's good practise not to overwrite arguments of functions. We can't use the original mol again in a later update because it's been overwritten

Reply:
This requires quite a significant change in code through all optimizers. Happy to write an issue and change this in the future.